compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -296,42 +296,21 @@ fn gather_explicit_predicates_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> ty::Gen
             bug!("unexpected {opaque_ty_node:?}")
         };
         debug!(?lifetimes);
-        for (arg, duplicate) in std::iter::zip(lifetimes, ast_generics.params) {
-            let hir::GenericArg::Lifetime(arg) = arg else { bug!() };
-            let orig_region = icx.astconv().ast_region_to_region(&arg, None);
-            if !matches!(orig_region.kind(), ty::ReEarlyBound(..)) {
-                // Only early-bound regions can point to the original generic parameter.
-                continue;
-            }
 
-            let hir::GenericParamKind::Lifetime { .. } = duplicate.kind else { continue };
-            let dup_def = duplicate.def_id.to_def_id();
-
-            let Some(dup_index) = generics.param_def_id_to_index(tcx, dup_def) else { bug!() };
+        let bidir_outlives = std::iter::zip(
+            lifetimes,
+            ast_generics
+                .params
+                .iter()
+                .filter(|param| matches!(param.kind, hir::GenericParamKind::Lifetime { .. }))
+                .map(|param| (param.def_id, param.name.ident().name, param.span)),
+        )
+        .map(|(arg, dup)| {
+            let hir::GenericArg::Lifetime(arg) = arg else { bug!() };
+            (**arg, dup)
+        });
 
-            let dup_region = ty::Region::new_early_bound(
-                tcx,
-                ty::EarlyBoundRegion {
-                    def_id: dup_def,
-                    index: dup_index,
-                    name: duplicate.name.ident().name,
-                },
-            );
-            predicates.push((
-                ty::Binder::dummy(ty::PredicateKind::Clause(ty::ClauseKind::RegionOutlives(
-                    ty::OutlivesPredicate(orig_region, dup_region),
-                )))
-                .to_predicate(icx.tcx),
-                duplicate.span,
-            ));
-            predicates.push((
-                ty::Binder::dummy(ty::PredicateKind::Clause(ty::ClauseKind::RegionOutlives(
-                    ty::OutlivesPredicate(dup_region, orig_region),
-                )))
-                .to_predicate(icx.tcx),
-                duplicate.span,
-            ));
-        }
+        predicates.extend(bidirectional_lifetime_predicates(tcx, def_id, bidir_outlives));
```

Comment:
We can also pass `&mut predicates` down instead of creating a new `Vec` but same reasoning as in https://github.com/rust-lang/rust/pull/112682/commits/e449766288c70225ba674d3b9e7b2230ea203f34#r1240243498, I guess this is probably optimized away?.

Reply:
You should just pass it down.

Reply:
Also,  you should probably only do any of the logic above if `in_trait` is true, right?

Reply:
oh, nevermind, you just extracted this function to use elsewhere.